### ISSUE TITLE
Any reason why not using the latest tensorflow and keras?

### ISSUE THREAD

**Issue by johnlockejrr** [qurator-spk/sbb_pixelwise_segmentation#22]
Is there any reason not using latest `tensorflow` and `keras`? Is very complicated to install into an environment (docker or WSL 2) with the old CUDA 11.8 and cuDNN 8.6. I refactored the scripts to use the new `tensorflow` (`2.19.0`) and seems to work so far for training.

![Image](https://github.com/user-attachments/assets/0422d9d6-959c-409d-aace-79819c2468c2)

**Comment by cneud** [comment]
Very interesting, thanks! 

TBH the main reason preventing us from upgrading is the switch from Keras v2 to Keras v3 API since TF 2.13. Apart from upgrading the codebase to use Keras v3 throughout, we may even have to either retrain models we previously trained (and keep on using productively) with Keras v3 and/or verify that there are no regressions or negative effects on output quality when switching to Keras v3 - which is cumbersome. 

Furthermore, as of Tensorflow 2.16, numpy 1.24 is required, whereas we still use numpy <1.24, since there are many expired deprecations in 1.24 that would require updating our codebase. 

While such updates would be very desirable, currently our work on Eynollah is funded through a project that is more focussed on improving methods and models, so that code maintenance and updating dependencies take a lower priority (since it still works). 

But we want to and will work more on maintenance aspects in the future, since we have to anyway to keep up with newer Python versions (cf. https://github.com/qurator-spk/eynollah/issues/122). Obviously any contributions or test that can help us towards this goal (like your test above) are especially welcome ;)

**Comment by johnlockejrr** [comment]
You are very right! I have ported `sbb_pixelwise_segmentation` to Keras v3 and I found many incompatibilities but it works. I have successfully trained a text line segmentation model and works for predicting and also with the old `Eynollah` models. The only major problem is the Keras v3 won't allow saving a model with `model.save()` without .keras or .h5 (so won't be backward compatible), you can do that with `model.export()` but will be useless if you want to finetune it later. Anyway, is achievable. With `Eynollah` is a little more work to do (the code is larger) but doable also (still working on it).

**Comment by johnlockejrr** [comment]
By the way, instead of upgrading to `tensorflow >2.13` and `Keras v3` why not going with `pythorch`? :)
Easier to maintain, more support, much more compatibility etc. And pytorch supports very well `resnet50`.

**Comment by johnlockejrr** [comment]
> We already use PyTorch in some other projects, also [timm](https://github.com/huggingface/pytorch-image-models) lures. Switching could provide an opportunity to revisit GPU code and make it more efficient. Certainly an option to consider, but frankly not something on our TODO list for at least this year.

I'm doing a test with `pytorch`, I'm not 100% sure that will work. Anyway, some training happens, the loss lowers and accuracy improves, fingers crossed.

![Image](https://github.com/user-attachments/assets/cd86c8ad-ecd9-4681-a9ac-0f13a17f67f3)

**Comment by johnlockejrr** [comment]
My model trained on pytorch:

![Image](https://github.com/user-attachments/assets/6a4e1bfe-70d3-4ce9-bbe2-c620208af34b)

**Comment by johnlockejrr** [comment]
The `pytorch` port is a little faster than `tensorflow` in prediction (also in training):

![Image](https://github.com/user-attachments/assets/1d219371-989c-47f2-9960-b239ecf4c540)

![Image](https://github.com/user-attachments/assets/ee25a64a-2b0f-44e9-a9a4-0b1e6c827f2b)

And the results comparable (left `pytorch`, right `tensorflow`):

![Image](https://github.com/user-attachments/assets/c9f06268-a8d0-4fb3-97da-9cd241d2ae42)
